### ISSUE TITLE
STF issues: using table key as Slice fails, syntax, etc.

### ISSUE THREAD

**Issue by hesingh** [p4lang/p4c#1906]
1. STF is a simple framework and thus I am not surprised, this line of code in my STF file fails.

```
STF Command: add lb_tbl meta.meta.lb_hash[19:12]:0x0 ingress.set_egress_spec(spec:11)
OrderedDict([(u'meta.meta.lb_hash[19:12]', u'exact')])
Exception  Unexpected key field meta.meta.lb_hash[19
```
`add lb_tbl meta.meta.lb_hash[19:12]:0x0 ingress.set_egress_spec(spec:11)`

What would it take to extend STF to support bit Slice? 

2. Also, why does behavioral-model "RuntimeCmd" use ```table_add <table name> <action name> <match fields> => <action parameters> [priority]``` but STF uses a different cmd and syntax shown above?  It would have been good for STF and RuntimeCmd to use the same cmd and syntax.

3. I think STF supports "packet 0" and "expect 2" which means packet ingressed at port 0 and egressed at port 2.  It would be good to add a P4 program to p4c repo to show how this is done in P4 code.  I have a rough idea which is to get_egress_port in the egress control.  Most STF tests I have seen use port 0 for both `packet` and `expect`

**Comment by hesingh** [comment]
I see dead code in the bmv2 backend related to a table key, if the key uses a bit Slice.  The dead code starts at the line below. 

https://github.com/p4lang/p4c/blob/master/backends/bmv2/common/control.cpp#L345

I am using a table key which is a bit Slice, but the code above did not get hit during testing.

For this issue, I have to make sure, the bmv2 backend is good to support a table key as bit Slice.   I don't think testdata/p4_16_sample/ includes any bmv2 test with table key as bit Slice. 

Thereafter, I can look at the behavioral-model for any code changes.

**Comment by ChrisDodd** [comment]
What would such a slice mean?  When you specify a table entry, you must specify the ENTIRE key required by the table -- specifying only part of the key makes no sense?

Most STF tests use expect 2 or 3 for output as until (relatively) recently, bmv2 would not output packets on port 0 properly (something to do with internal drop handling?).  So only a few recent STF tests use expect 0

**Comment by jafingerhut** [comment]
P4_16 lets you specify a bit slice expression (and many other kinds of arithmetic expressions) as a key of a table, it appears that the existing STF test infrastructure does not let you refer to this key as the name of the field of which a slice is being taken, nor does it let you specify the slice expression, either.

That isn't a problem with P4_16, only (potentially) a problem with STF generality.

Hemant, you can use annotations `@name("different_name")` _after_ the match_kind (e.g. exact) in a table key definition to specify a different name to be used for that table key, whether it is an arithmetic expression or not, to work around this problem, although in experimenting with that I have found a case where the JSON file produced has a syntax error because it tries to include `@name("different_name")` inside of a quoted string, without escaping or changing the inner double quote characters.  I will create a separate issue for that.

**Comment by ChrisDodd** [comment]
When you use an expression as a table key (a slice or otherwise), it requires a @name annotation to set the name for that key (part) in the API.  That API name is what is required in the STF test.  When you have a "simple" key (just a field), it implicitly gets an @name annotation to set the API name to use, which is the name STF uses to refer to that key.

So this isn't problem with STF specifically, its just the way API specification works.

**Comment by jafingerhut** [comment]
Well, at least in my quick tests today, if the key expression is something like "field - 10", I get an error message from the compiler saying that a @name annotation is required.

If the expression is "field[5:2]" or "field & 0x0101", there is no such error message, and the name of the field becomes that expression in the BMv2 JSON file.

**Comment by hesingh** [comment]
@jafingerhut I agree a bit Slice must be allowed to be a table key.   It is interesting behavior by p4c, on seeing a bit Slice for table key, p4c does not say "key is complex, please add @name annotation".   I am using p4c-bm2-ss.   I did add the name and see the json output change.  But the behavioral-model error still persists. 

@ChrisDodd thanks for all the details.  I changed my stf file to use the @name name and STF still  complains.  This is changed code: ` meta.meta.lb_hash[19:12] : exact @name("lb_key");`

I ran this test: `$ bmv2/testdata/p4_16_samples/switch_foo-bmv2.p4.test` 

The error is:

```
STF Command: add lb_tbl meta.meta.lb_key:0x0 ingress.set_egress_spec(spec:11)
OrderedDict([(u'lb_key', u'exact')])
Exception  Unexpected key field meta.meta.lb_key
```

We do have to address the dead code in the bmv2 backend as well.

**Comment by ChrisDodd** [comment]
With @name("lb_key"), you need to use `add lb_tbl lb_key:0x0 ingress`... -- the STF command needs to match the @name.

At one point there was a proposed extension that moderately simple key expressions (like slices, or masks) with no @name annotation would get a @name from the field name, but I thought that was rejected.  But apparently p4c accepts it anyways?

In any case, it makes no sense to add a special case to STF that would not be accepted by a 'standard' API.

**Comment by hesingh** [comment]
Got it - Chris!  My colleague Alex also emailed me the same suggestion - thanks.

**Comment by mbudiu-vmw** [comment]
Regarding STF files: we actually have two STF file parsers. One is in tools, and the other one is hardwired in the run-*.py scripts. The one in the tools folder is much more complete. We have an issue open #1489 to migrate the python scripts to use the tools version, but obviously this has been a low priority.

For all expressions that can appear in P4-14 programs we have to generate automatically `@name` annotations, since there is no way users can write them. For some P4-16 expressions we also generate name annotations automatically. I am not sure whether either of the STF parsers supports the full range of annotations that are automatically generated. We should have some tests.

**Comment by hesingh** [comment]
@mbudiu-vmw thanks for the update.

**Comment by jafingerhut** [comment]
Is there anything remaining from the discussion above that needs to be done before closing this issue?  As Mihai mentioned, there is already an open issue for the semi-big task of updating the STF test infrastructure of p4c to enable more kinds of tests to be written, but I don't see anything that looks like a bug, as much as field name limitations in the existing (and maybe also future) STF tests, which can be worked around by using @name annotations on fields.

**Comment by hesingh** [comment]
Ok, we can close this one - thanks.  I just did.
